src/common/opencl.c

```diff
@@ -2438,38 +2441,96 @@ void dt_opencl_memory_statistics(int devid, cl_mem mem, dt_opencl_memory_t actio
                                       (float)darktable.opencl->dev[devid].memory_in_use/(1024*1024));
 }
 
-/** check if image size fit into limits given by OpenCL runtime */
-int dt_opencl_image_fits_device(const int devid, const size_t width, const size_t height, const unsigned bpp,
-                                const float factor, const size_t overhead)
+cl_ulong dt_opencl_get_device_available(const int devid)
 {
-  static float headroom = -1.0f;
+  if(!darktable.opencl->inited || devid < 0) return 0;
+  const size_t calculated = darktable.opencl->dev[devid].available_mem;
+  return (darktable.unmuted & DT_DEBUG_LOWMEM) ? MIN(calculated, 512lu * 1024lu * 1024lu) : calculated ;
+}
 
-  if(!darktable.opencl->inited || devid < 0) return FALSE;
+cl_ulong dt_opencl_get_device_memalloc(const int devid)
+{
+  if(!darktable.opencl->inited || devid < 0) return 0;
+  return darktable.opencl->dev[devid].max_mem_alloc;
+}
 
-  /* first time run */
-  if(headroom < 0.0f)
-  {
-    headroom = dt_conf_get_float("opencl_memory_headroom") * 1024.0f * 1024.0f;
+/* As there is no portable way to get the unused memory of a cl device we check this by allocating
+   buffers until there is no room left.
+   clCreateBuffer does not necessarily tell an error condition if there is no memory available (this
+   is according to standard and certainly tru for nvidia), to make sure the buffer is really allocated
+   in graphics mem we force a small memory access.
```

Comment:
Wouldn't this approach interfere with other applications using GPU memory? I mean they might get OOM errors if they try to allocate GPU buffer while dt is doing this check. Depending on the frequency of the check probability of this problem can be quite high. Also, doesn't it interfere with dt pipe itself, or is this check done only when we're sure no dt pipe is running?
Point is that I'm afraid such approach will cause very strange and hard-to-debug problems on the user's system. It would be better to find some non-interfering option to get free GPU memory...

Reply:
or add a new option like headroom to define the amount of memory to leave free for future allocations. So darktable works out how much memory is currently used, adds an amount defined in the preference (100Mb by default, say) and uses whatever is left over, leaving 100Mb free for other applications to allocate.

Reply:
the problem is with calculation process, not about pipeline run itself. I think headroom still exist, see other comments in this PR.

Reply:
Bad flushing-red head here on my side :-( The proposed approach in this pr is indeed leaving dt as the "bad guy" not respecting shared resources. 

I will read through docs if there is another way to get available mem, otherwise will do a check just inside `dt_opencl_image_fits_device()` as a safe way.